### PR TITLE
第二批单测规范化建设

### PR DESCRIPTION
**Pull request by Xuxuanang**
### PR Docs
算子单测规范化建设

### PR APIs

torch.autograd.function.FunctionCtx
torch.cuda.BFloat16Tensor
torch.cuda.Stream
torch.cuda.amp.GradScaler
torch.cuda.current_device
torch.cuda.initial_seed
torch.cuda.is_bf16_supported
torch.einsum
torch.frexp
torch.get_default_dtype
torch.nn.DataParallel
torch.nn.GRU
torch.nn.LSTM
torch.nn.RNN

```

### CODE REVIEW

**Comment by zhwesky2010** [comment]
@Xuxuanang  pre-commit代码风格修改一下

File: tests/test_cuda_amp_GradScaler.py [hunk 1]
```diff
@@ -14,39 +14,73 @@
...
+            assert np.allclose(
+                pytorch_numpy, paddle_numpy, rtol=rtol, atol=atol
+            ), "API ({}): paddle result has diff with pytorch result".format(name)
 
-def _test_case_1():
+
+obj = cuda_amp_GradScalerAPIBase("torch.cuda.amp.GradScaler")
+
+
+def test_case_1():
```
Comment (zhwesky2010) [inline tests/test_cuda_amp_GradScaler.py:51]: 加装饰器跳过吧，装饰器是规范写法：
```
@pytest.mark.skipif(
    condition=not paddle.device.is_compiled_with_cuda(),
    reason="can only run on paddle with CUDA",
)
```

File: tests/test_cuda_amp_GradScaler.py [hunk 2]
```diff
@@ -14,39 +14,73 @@
...
-        result = scaled
+        if not torch.cuda.is_available():
+            result = torch.Tensor([1])
+        else:
+            scaler = torch.cuda.amp.GradScaler()
+            x = torch.tensor([[[-1.3020, -0.1005,  0.5766,  0.6351, -0.8893,  0.0253, -0.1756, 1.2913],
+                                [-0.8833, -0.1369, -0.0168, -0.5409, -0.1511, -0.1240, -1.1870, -1.8816]]])
+            with torch.cuda.amp.autocast():
+                loss = torch.mean(x*x).to('cpu')
+            scaled = scaler.scale(loss).cpu()  # scale the loss
```
Comment (zhwesky2010) [inline tests/test_cuda_amp_GradScaler.py:63]: scaled这个变量多余？或者直接：`obj.run(pytorch_code, ["scaled"])`

File: tests/test_cuda_amp_GradScaler.py [hunk 3]
```diff
@@ -14,39 +14,73 @@
...
+            result = scaled
         """
     )
     obj.run(pytorch_code, ["result"])
 
 
-def _test_case_2():
+def test_case_2():
```
Comment (zhwesky2010) [inline tests/test_cuda_amp_GradScaler.py:70]: 同上

File: tests/test_cuda_current_device.py [hunk 4]
```diff
@@ -19,12 +19,15 @@
...
     pytorch_code = textwrap.dedent(
         """
         import torch
-        torch.cuda.current_device()
-        result = True
+        if not torch.cuda.is_available():
+            result =  1
+        else:
+            torch.cuda.current_device()
+            result = True
```
Comment (zhwesky2010) [inline tests/test_cuda_current_device.py:30]: 1. 统一用装饰器规范写法，不要这种判断分支
2. 这里的单测写的不规范，result并没有得到真实的测试，你可以修改compare函数来适配这个API的返回结果

File: tests/test_cuda_initial_seed.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+from apibase import APIBase
+
+obj = APIBase("torch.cuda.initial_seed")
+
+
+def test_case_1():
+    pytorch_code = textwrap.dedent(
+        """
+        import torch
+        if not torch.cuda.is_available():
```
Comment (zhwesky2010) [inline tests/test_cuda_initial_seed.py:26]: 1. 统一用装饰器规范写法，不要这种判断分支
2. 这里的单测写的不规范，result并没有得到真实的测试，你可以修改compare函数来适配这个API的返回结果

File: tests/test_cuda_is_bf16_supported.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+import textwrap
+
+from apibase import APIBase
+
+obj = APIBase("torch.cuda.is_bf16_supported")
+
+
+def test_case_1():
+    pytorch_code = textwrap.dedent(
```
Comment (zhwesky2010) [inline tests/test_cuda_is_bf16_supported.py:23]: 1. 统一用装饰器规范写法，不要这种判断分支
2. 这里的单测写的不规范，result并没有得到真实的测试，你可以修改compare函数来适配这个API的返回结果

File: tests/test_frexp.py [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    )
+    obj.run(pytorch_code, ["result"])
+
+
+def test_case_2():
+    pytorch_code = textwrap.dedent(
+        """
+        import torch
+        x = torch.tensor([[128.0, 64.0], [-32.0, 16.0]])
+        result, ex = torch.frexp(x)
```
Comment (zhwesky2010) [inline tests/test_frexp.py:38]: ex没有测

File: tests/test_frexp.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+obj = APIBase("torch.frexp")
+
+
+def test_case_1():
+    pytorch_code = textwrap.dedent(
+        """
+        import torch
+        x = torch.tensor([10.0, -2.5, 0.0, 3.14])
+        result, exponent = torch.frexp(x)
```
Comment (zhwesky2010) [inline tests/test_frexp.py:27]: exponent没有测，如果转换策略有问题，对不齐第二个返回值，就需要修改转换策略

File: tests/test_nn_DataParallel.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+        check_value=True,
+        check_dtype=True,
+        check_stop_gradient=True,
+        rtol=1.0e-6,
+        atol=0.0,
+    ):
+        (
+            pytorch_numpy,
+            paddle_numpy,
+        ) = pytorch_result.cpu().detach().numpy(), paddle_result.numpy(False)
```
Comment (zhwesky2010) [inline tests/test_nn_DataParallel.py:36]: 这个重写，并没有修改compare函数的实质逻辑？重写意义是？

File: tests/test_nn_DataParallel.py [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+
+            def forward(self, x):
+                return self.fc(x)
+
+        x = torch.tensor([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]])
+        model = nn.DataParallel(SimpleModel(), device_ids=[0], output_device=0)
+        result = model(x)
+        """
+    )
+    obj.run(pytorch_code, ["result"],check_value=False)
```
Comment (zhwesky2010) [inline tests/test_nn_DataParallel.py:122]: 常数初始化，还需要 check_value=False吗，这里为何对不齐结果

File: tests/test_nn_GRU.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+
+def test_case_5():
+    pytorch_code = textwrap.dedent(
+        """
+        import torch
+        import torch.nn as nn
+        class SimpleRNNModel(nn.Module):
+            def __init__(self):
+                super(SimpleRNNModel, self).__init__()
+                self.gru = nn.GRU(10,3,2,True,False,0.7,False)
```
Comment (zhwesky2010) [inline tests/test_nn_GRU.py:123]: 注重一下代码规范：相邻参数空一格
```
nn.GRU(10, 3, 2, True, False, 0.7, False)
```

File: tests/test_nn_LSTM.py [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+
+
+def test_case_1():
+    pytorch_code = textwrap.dedent(
+        """
+        import torch
+        class SimpleRNNModel(torch.nn.Module):
+            def __init__(self):
+                super(SimpleRNNModel, self).__init__()
+                self.lstm = torch.nn.LSTM(input_size=10, hidden_size=20, num_layers=2, bias=True, batch_first=True, dropout=0.5,bidirectional=True, proj_size = 0)
```
Comment (zhwesky2010) [inline tests/test_nn_LSTM.py:30]: proj_size 不等于 0时测一下结果

Comment (zhwesky2010) [inline tests/test_nn_LSTM.py:30]: bias 为True/False都测一下

File: tests/test_nn_RNN.py [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+def test_case_1():
+    pytorch_code = textwrap.dedent(
+        """
+        import torch
+        import torch.nn as nn
+
+        class SimpleRNNModel(nn.Module):
+            def __init__(self):
+                super(SimpleRNNModel, self).__init__()
+                self.rnn = nn.RNN(input_size=10, hidden_size=20, num_layers=2, nonlinearity='tanh',bias=True,batch_first=True, dropout=0.5,
```
Comment (zhwesky2010) [inline tests/test_nn_RNN.py:32]: bias 为True/False都测一下

File: tests/test_nn_LSTM.py [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+
+
+import textwrap
+
+from apibase import APIBase
+
+obj = APIBase("torch.nn.LSTM")
+
+
+def test_case_1():
```
Comment (zhwesky2010) [inline tests/test_nn_LSTM.py:23]: LSTM和RNN这两个单测还能多加一下case，这两个API的参数很多，参数bool取值，True/False多种组合形式都需覆盖一下，尽可能不留测试死角

File: tests/test_nn_DataParallel.py (hunk 9 above)
Comment (Xuxuanang) [reply tests/test_nn_DataParallel.py:36]: 这个是因为不能把gpu上面的张量直接转换为numpy数组比较大小，需要先转换到cpu上面

Comment (zhwesky2010) [reply tests/test_nn_DataParallel.py:36]: > 这个是因为不能把gpu上面的张量直接转换为numpy数组比较大小，需要先转换到cpu上面

原来不也是这个逻辑吗？

infoflow 2024-08-07 21-31-13

File: paconvert/api_matcher.py [hunk 15]
```diff
@@ -4610,3 +4613,9 @@ def get_paddle_class_nodes(self, func, args, kwargs):
 
         code = "{}({})".format(self.get_paddle_api(), self.kwargs_to_str(kwargs))
         return ast.parse(code).body
+
+class FrexpMatcher(BaseMatcher):
+    def generate_code(self, kwargs):
+        code = "( paddle.frexp(x = {})[0], paddle.frexp(x = x)[1].astype('int32'))".format(kwargs["input"])
```
Comment (zhwesky2010) [inline paconvert/api_matcher.py:4619]: 这个是因为paddle返回的第二个参数是int64吗，这个差异的影响很小，就按check_dtype=False来测吧。

**Comment by zhwesky2010** [comment]
@Xuxuanang 你这个CI还没通过，需要CI流水线全部通过才能合入

File: tests/test_nn_DataParallel.py (hunk 9 above)
Comment (Xuxuanang) [reply tests/test_nn_DataParallel.py:36]: 原来会进入到这个分支
屏幕截图 2024-08-08 002803

**Comment by zhwesky2010** [comment]
@Xuxuanang 处理一下冲突

**Review by zhwesky2010 — APPROVED** [review]
